source/common/grpc/async_client_impl.cc

```diff
@@ -77,7 +77,7 @@ void AsyncStreamImpl::initialize(bool buffer_body_for_retry) {
 }
 
 void AsyncStreamImpl::onHeaders(Http::HeaderMapPtr&& headers, bool end_stream) {
-  ASSERT(!remote_closed_);
+  callbacks_.onReceiveInitialMetadata(std::move(headers));
   const auto http_response_status = Http::Utility::getResponseStatus(*headers);
```

Comment:
I think this is a illegal use of `headers` as you moved it above.

Reply:
I agree, surprised the compiler/sanitizers didn't catch that. Will fix.